File: app/src/main/java/com/recursia/popularmovies/domain/MoviesRepository.java
package com.recursia.popularmovies.domain;

import com.recursia.popularmovies.domain.models.Movie;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Single;

public interface MoviesRepository {

    Single<List<Movie>> discoverMovies(String sortBy, int page, int voteCount, String language);

    Flowable<List<Movie>> getAllFavoriteMovies();

    Single<Movie> getMovieById(int movieId, String language);

    Single<List<Movie>> getMoviesByQuery(String query, int page, String language);

    Single<List<Movie>> getMovieRecommendations(int movieId, int page, String language);

    Completable makeFavoriteMovie(Movie movie);

    Completable removeFavoriteMovie(Movie movie);

}

File: app/src/main/java/com/recursia/popularmovies/domain/MoviesRepository.kt
package com.recursia.popularmovies.domain

import com.recursia.popularmovies.domain.models.Genre
import com.recursia.popularmovies.domain.models.Movie
import com.recursia.popularmovies.domain.models.enums.Category
import io.reactivex.Single

interface MoviesRepository {

    fun getMovieById(movieId: Int, language: String): Single<Movie>

    fun getMoviesByQuery(query: String, page: Int, language: String): Single<List<Movie>>

    fun getMovieRecommendations(movieId: Int, page: Int, language: String): Single<List<Movie>>

    fun getMoviesWithCategory(category: Category, page: Int, language: String): Single<List<Movie>>

    fun getGenreMovies(genre: Genre, page: Int, language: String): Single<List<Movie>>

    fun getGenres(language: String): Single<List<Genre>>
}
